Kestrel response header encoding

Tratcher opened the pull request:
Fixes #4727

I have all of the plumbing in place with tests written and passing, but there's plenty of room to clean it up.
- Added Http3HeadersEncoder for consistency with HTTP/2.
- Response.Headers validation has been relaxed if you're using a non-default encoder. Control characters are still blocked, but non-ascii characters will be allowed.
  - If the custom encoder decides to throw an exception when serializing the response rather than using a default replacement character, that will cause the connection to be aborted (HTTP/1 & 2, not sure about 3) to avoid inconsistent states like out of sync dynamic encoding. Applications can filter response headers in middleware/OnResponseStarting to avoid this.
- Fixed a few spots where the request header encoding wasn't null checked.

TODO:
- Sync HPack changes back to runtime

cc: @benaadams 


ResponseHeadersWritingBenchmark (HTTP/1)

Before:
| Method |                 Type |     Mean |     Error |    StdDev |   Median |      Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |--------------------- |---------:|----------:|----------:|---------:|----------:|------:|------:|------:|----------:|
| Output | TechEmpowerPlaintext | 1.403 us | 0.0321 us | 0.0874 us | 1.400 us | 712,510.4 |     - |     - |     - |     144 B |
| Output |     PlaintextChunked | 1.241 us | 0.0287 us | 0.0786 us | 1.200 us | 805,555.6 |     - |     - |     - |     144 B |
| Output |  PlaintextWithCookie | 1.617 us | 0.0609 us | 0.1748 us | 1.600 us | 618,489.6 |     - |     - |     - |     144 B |
| Output | Plain(...)ookie [26] | 1.635 us | 0.0367 us | 0.0906 us | 1.600 us | 611,724.7 |     - |     - |     - |     144 B |
| Output |           LiveAspNet | 3.493 us | 0.0687 us | 0.1257 us | 3.500 us | 286,298.6 |     - |     - |     - |     144 B |

After, default encoder:
| Method |                 Type |     Mean |     Error |    StdDev |   Median |      Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |--------------------- |---------:|----------:|----------:|---------:|----------:|------:|------:|------:|----------:|
| Output | TechEmpowerPlaintext | 1.563 us | 0.0352 us | 0.0720 us | 1.600 us | 639,899.6 |     - |     - |     - |     144 B |
| Output |     PlaintextChunked | 1.468 us | 0.0332 us | 0.0476 us | 1.500 us | 681,265.2 |     - |     - |     - |     144 B |
| Output |  PlaintextWithCookie | 1.934 us | 0.0467 us | 0.1332 us | 1.900 us | 517,051.7 |     - |     - |     - |     144 B |
| Output | Plain(...)ookie [26] | 2.051 us | 0.0397 us | 0.0736 us | 2.100 us | 487,528.3 |     - |     - |     - |     144 B |
| Output |           LiveAspNet | 3.661 us | 0.0780 us | 0.2162 us | 3.700 us | 273,173.7 |     - |     - |     - |     144 B |

After, utf8 encoder:
| Method |                 Type |     Mean |     Error |    StdDev |      Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |--------------------- |---------:|----------:|----------:|----------:|------:|------:|------:|----------:|
| Output | TechEmpowerPlaintext | 1.806 us | 0.0482 us | 0.1344 us | 553,846.2 |     - |     - |     - |     144 B |
| Output |     PlaintextChunked | 1.810 us | 0.0545 us | 0.1573 us | 552,359.0 |     - |     - |     - |     144 B |
| Output |  PlaintextWithCookie | 2.118 us | 0.0493 us | 0.1390 us | 472,036.9 |     - |     - |     - |     144 B |
| Output | Plain(...)ookie [26] | 2.103 us | 0.0578 us | 0.1660 us | 475,475.5 |     - |     - |     - |     144 B |
| Output |           LiveAspNet | 3.877 us | 0.0992 us | 0.2781 us | 257,936.5 |     - |     - |     - |     144 B |

HPackHeaderWriterBenchmark (Http/2)

Before:
|                            Method |     Mean |   Error |  StdDev |        Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------------------- |---------:|--------:|--------:|------------:|------:|------:|------:|----------:|
|   BeginEncodeHeaders_KnownHeaders | 636.2 ns | 1.89 ns | 1.68 ns | 1,571,833.8 |     - |     - |     - |         - |
| BeginEncodeHeaders_UnknownHeaders | 724.9 ns | 4.26 ns | 3.78 ns | 1,379,445.1 |     - |     - |     - |         - |

After:

|                                           Method |     Mean |   Error |  StdDev |        Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------------------------- |---------:|--------:|--------:|------------:|------:|------:|------:|----------:|
|                  BeginEncodeHeaders_KnownHeaders | 682.1 ns | 2.02 ns | 1.89 ns | 1,465,996.9 |     - |     - |     - |         - |
|   BeginEncodeHeaders_KnownHeaders_CustomEncoding | 816.1 ns | 2.87 ns | 2.69 ns | 1,225,306.0 |     - |     - |     - |         - |
|                BeginEncodeHeaders_UnknownHeaders | 773.0 ns | 2.75 ns | 2.29 ns | 1,293,619.5 |     - |     - |     - |         - |
| BeginEncodeHeaders_UnknownHeaders_CustomEncoding | 750.2 ns | 3.03 ns | 2.83 ns | 1,333,002.7 |     - |     - |     - |         - |


Crank Plaintext default encoding:

| application           | old                        | new                        |         |
| --------------------- | -------------------------- | -------------------------- | ------- |
| CPU Usage (%)         |                        100 |                         99 |  -1.00% |
| Cores usage (%)       |                      1,194 |                      1,193 |  -0.08% |
| Working Set (MB)      |                         72 |                         71 |  -1.39% |
| Private Memory (MB)   |                        395 |                        370 |  -6.33% |
| Build Time (ms)       |                      2,951 |                      2,993 |  +1.42% |
| Start Time (ms)       |                        162 |                        202 | +24.69% |
| Published Size (KB)   |                    106,170 |                    106,170 |   0.00% |
| .NET Core SDK Version | 6.0.100-preview.6.21276.12 | 6.0.100-preview.6.21276.12 |         |


| load                   | old        | new        |         |
| ---------------------- | ---------- | ---------- | ------- |
| CPU Usage (%)          |         42 |         43 |  +2.38% |
| Cores usage (%)        |        503 |        513 |  +1.99% |
| Working Set (MB)       |         38 |         38 |   0.00% |
| Private Memory (MB)    |        358 |        358 |   0.00% |
| Start Time (ms)        |          0 |          0 |         |
| First Request (ms)     |         97 |        149 | +53.61% |
| Requests/sec           |  2,410,464 |  2,342,838 |  -2.81% |
| Requests               | 36,397,228 | 35,375,101 |  -2.81% |
| Mean latency (ms)      |       1.17 |       1.14 |  -2.56% |
| Max latency (ms)       |      49.73 |      41.52 | -16.51% |
| Bad responses          |          0 |          0 |         |
| Socket errors          |          0 |          0 |         |
| Read throughput (MB/s) |     303.44 |     294.93 |  -2.80% |
| Latency 50th (ms)      |       0.95 |       0.97 |  +2.11% |
| Latency 75th (ms)      |       1.39 |       1.41 |  +1.44% |
| Latency 90th (ms)      |       1.79 |       1.80 |  +0.56% |
| Latency 99th (ms)      |       0.00 |       0.00 |         |

Tratcher committed:
TEMP: Add response header encoding callback

Tratcher committed:
Add trailers test

Tratcher committed:
HTTP/2 response headers test

Tratcher committed:
HTTP/1 custom header encoding

Tratcher committed:
Custom encoding for known headers

Tratcher committed:
Validate custom encodings

Tratcher committed:
Cleanup

Tratcher committed:
Http/2 hpack encoding

Tratcher committed:
More validation, HTTP/2 trailers tests

Tratcher committed:
Http/3

gfoidl commented on src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpCharacters.cs:
```diff
@@ -198,5 +199,23 @@ public static int IndexOfInvalidFieldValueChar(string s)
+
+        // Disallows control characters but allows extended characters > 0x7F
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int IndexOfInvalidFieldValueCharExtended(string s)
```
Unrelated to this PR, but are these methods hot?
My gut-feeling means there may be some room for perf-improvements, so is it worth to have a look?

Tratcher replied:
It's called multiple times per response and some values like Set-Cookie could be long, so yes, feel free to take a look.

gfoidl replied:
I gave this a quick try, see speedups (depending on the actual length) of 5 to 10.
Will create a PR once this PR is in.

gfoidl replied:
Wanted to wait for xplat-intrinsics to be there 😃 
--> https://github.com/dotnet/aspnetcore/pull/44041

stephentoub replied:
@gfoidl, what generalized IndexOfXx methods could we expose in runtime to avoid the need to custom code something here? Would https://github.com/dotnet/runtime/issues/68328 suffice?

gfoidl replied:
Hm, yes I think `IndexOfAny` for > 5 can do the job.
I didn't think about that, as my prototype dates back to June 2021 and then I didn't tidy up my personal todo with the new possibilities.

I have an idea on how that could work generalized and based on a generator to get the best of both worlds (vector supported vs. not supported). A bit later today (have to leave now) I'll sketch out a prototype of this and share my findings / thoughts in https://github.com/dotnet/runtime/issues/68328.

Tratcher commented on src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.cs:
```diff
@@ -260,24 +261,50 @@ IEnumerator IEnumerable.GetEnumerator()
             {
-                var invalid = HttpCharacters.IndexOfInvalidFieldValueChar(headerCharacters);
-                if (invalid >= 0)
+                // Only validate here if we're using the default encoding (ASCII). Otherwise we'll validate later when encoding.
```
```suggestion
```

Tratcher committed:
Add Http3HeadersEnumerator

Tratcher committed:
Efficient encoding

Tratcher committed:
Reduce validation on add, error handling

Tratcher committed:
Http/3 error handling and tests

Tratcher committed:
Cleanup TODOs

Tratcher committed:
Fixup request encoding

Tratcher committed:
Clean up generated spacing

Tratcher committed:
Formatting

Tratcher committed:
Count encoded length

Tratcher committed:
hpack benchmark improvements

Tratcher commented:
/AzurePipelines run AspNetCore-benchmarking-pr

gfoidl commented on src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpCharacters.cs:
```diff
@@ -182,6 +182,7 @@ public static int IndexOfInvalidTokenChar(ReadOnlySpan<byte> span)
             return -1;
         }
 
+        // Disallows control characters and anything more than 0x7E
```
```suggestion
        // Disallows control characters and anything more than 0x7F
```
?

Tratcher replied:
😁 7E is also a prohibited control character (Delete), but I will change the comment to be consistent.

gfoidl commented on src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpCharacters.cs:
```diff
@@ -198,5 +199,23 @@ public static int IndexOfInvalidFieldValueChar(string s)
+            for (var i = 0; i < s.Length; i++)
+            {
+                var c = s[i];
+                if (c < (uint)fieldValue.Length && !fieldValue[c])
```
Is this condition correct?

We want to disallow control characters (given by `fieldValue`), and allow characters > 0x7F.
Say we get 'A' (ASCII: 65), then it looks like `if (65 < 128 && ...)` which is true and the `i` is returned. Expected is `-1` as return value.

I see CI looks fine, so it's either me missing something or there's a test missing.

Tratcher replied:
I think it's right.
`if (c < (uint)fieldValue.Length && !fieldValue[c])` where c is `A` is `if (65 < 128 && !fieldValue[65])` where `fieldValue[65]` is `true` so `if (65 < 128 && !true)` so the if fails and the loop does not exit early.

gfoidl replied:
You're right.

> so it's either me missing something

It's me...`&&` and not `||` which my brain thought about. 
Thanks for clarification!

Tratcher committed:
PR feedback

JunTaoLuo commented on src/Shared/runtime/Http2/Hpack/HeaderField.cs:
```diff
@@ -36,7 +36,7 @@ public override string ToString()
             if (Name != null)
             {
-                return Encoding.ASCII.GetString(Name) + ": " + Encoding.ASCII.GetString(Value);
+                return Encoding.Latin1.GetString(Name) + ": " + Encoding.Latin1.GetString(Value);
```
Why did this encoding change?

halter73 replied:
This is just for debugging, right? Since we cannot know that's what's intended and this could garble some encodings like UTF-8, I think we should revert it.

Tratcher replied:
It is just for debugging. With ASCII you'd only get replacement characters and lose information, at least with Latin1 you'd get representative characters that could be compared to other values. Are you sure you want to revert it?

halter73 replied:
If you think it's useful, I don't feel that strongly.

JunTaoLuo commented on src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestHeaders.cs:
```diff
@@ -97,16 +97,16 @@ private void AppendContentLength(ReadOnlySpan<byte> value)
             // long.MaxValue = 9223372036854775807 (19 chars)
-            Span<char> decodedChars = stackalloc char[20];
-            var numChars = customEncoding!.GetChars(value, decodedChars);
+            Span<char> decodedChars = stackalloc char[32]; // power of two
```
Curious, why is using a power of two better?

gfoidl replied:
I don't know the exact rationale of the JIT here, but by [looking at the codegen](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4a9JqwAqMBBg3VNAZiakGAMQgQGAbxoMnTM2AAW2KAwCCACgCUjs4O1M5hDLJi2FwAPO6eAHwMuFFcDAC8yRjYYADW2JKSEGAM8VAA2qR0ALpG4c7EAOzJqeU1dc4AvkFOPS6lHl4AQgF9IfVOkdFxg0kp0RlZOfmFxQOe5SaktX1hTS3RbTuhXX195bK5vGIAMsUFuACSXLwY1X3EroM+5KMnTuMJlNYmU5qlFrhsnkCkUSmVKu1dg1mvMuEcOk5uv8GOdLtc7mAHs9Xu9sZ91sNfoFsYD6sCZokDmlMpDljC1vCtscJkwUa1EdisZ0gA===) a power of two is better.

Here I believe an optimization comes into play, that treats small stack alloc sizes as normal locals.
So maybe that's also the reason why SkipLocalsInit has no effect when 20 is used size. With a power of two, the zeroing isn't done (as intentend). In the sharplab linked above you can try to increase the allocation size (e.g. to 120 and 128) and see that there's a difference between small sizes and larger ones.

halter73 replied:
Interesting. I don't think that `// power of two` explains this sufficiently though. Playing with the SharpLab, it looks like the extra, non-SkipLocalsInit affected stackalloced spans happen betweens 33-48 bytes which 20 chars happens to fall in-between. @gfoidl Do you know if there's a runtime issue tracking this?

I think we should only ask for what we need and let the JIT optimize how it sees fit.

gfoidl replied:
> 33-48 bytes

The sharplab is for `char`, so 66-96 bytes. I can't find any relevant issue in dotnet/runtime, but I remember there were some changes in the way how stack space is zeroed with various threshold and 96 bytes being one of them (IIRC).

For a proper reasoning what's going on here we should ask JIT-team. Maybe it's a bug, or it's by intention.
At least it's a bit surprising, especially that SkipLocalsInit hasn't the effect when not a power of two and small allocation sizes.

> we should only ask for what we need and let the JIT optimize how it sees fit.

The span is sliced a bit later to the correct size anyway, so the size is in excess at time of allocation. But your view is more "natural" to rounding up to a power of two.
BTW: In dotnet/runtime almost always a power of two is used for stackalloc.

Tratcher replied:
I'll revert this for now since it's unrelated to this PR. I recommend following up with the JIT folks to clarify the recommendations here.

halter73 replied:
> The sharplab is for char, so 66-96 bytes. 

I was testing with `stackalloc byte[N]`.

gfoidl replied:
Filed https://github.com/dotnet/runtime/issues/55001 to track this.

JunTaoLuo commented on src/Servers/Kestrel/Core/src/Internal/Http/HttpResponseHeaders.cs:
```diff
@@ -18,6 +19,13 @@ internal sealed partial class HttpResponseHeaders : HttpHeaders
 
+        public Func<string, Encoding?> EncodingSelector { get; set; }
+
+        public HttpResponseHeaders(Func<string, Encoding?>? encodingSelector = null)
```
Instead of nullable, why not set the default value to KestrelServerOptions.DefaultHeaderEncodingSelector? I think I see this in a few places.

JunTaoLuo replied:
Nvm it's not possible

JunTaoLuo reviewed (commented):
First batch

JunTaoLuo commented on src/Servers/Kestrel/Core/src/Internal/Http3/Http3HeadersEnumerator.cs:
```diff
@@ -0,0 +1,152 @@
+
+        internal static int GetResponseHeaderStaticTableId(KnownHeaderType responseHeaderType)
+        {
+            // Not Implemented
```
When will we need to do this? Do we need to file a followup issue?

Tratcher replied:
https://github.com/dotnet/aspnetcore/issues/33980

JunTaoLuo commented on src/Shared/ServerInfrastructure/BufferExtensions.cs:
```diff
@@ -199,32 +199,60 @@ private static void WriteNumericMultiWrite(ref this BufferWriter<PipeWriter> buf
+            var minBufferSize = encoding.GetMaxByteCount(1);
+            var completed = false;
+
+            // This may be a bug, but encoder.Convert returns completed = true for UTF7 too early.
```
Is there an issue filed for UTF7 so they can investigate?

Tratcher replied:
https://github.com/dotnet/runtime/issues/414

JunTaoLuo commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -189,7 +189,7 @@ public void WriteResponseHeaders(int streamId, int statusCode, Http2HeadersFrame
                     FinishWritingHeaders(streamId, payloadLength, done);
                 }
-                catch (HPackEncodingException hex)
+                catch (Exception hex) // Any exception from the HPack encoder can leave the dynamic table in a corrupt state.
```
```suggestion
                catch (Exception ex) // Any exception from the HPack encoder can leave the dynamic table in a corrupt state.
```

halter73 replied:
Maybe point out that custom encoders are a reason we cannot be more certain about what Exception types might be thrown?

JunTaoLuo commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:
```diff
@@ -215,7 +215,7 @@ public ValueTask<FlushResult> WriteResponseTrailersAsync(int streamId, HttpRespo
                     FinishWritingHeaders(streamId, payloadLength, done);
                 }
-                catch (HPackEncodingException hex)
+                catch (Exception hex)  // Any exception from the HPack encoder can leave the dynamic table in a corrupt state.
```
```suggestion
                catch (Exception ex)  // Any exception from the HPack encoder can leave the dynamic table in a corrupt state.
```

gfoidl commented on src/Servers/Kestrel/shared/test/StreamBackedTestConnection.cs:
```diff
@@ -20,10 +20,10 @@ public abstract class StreamBackedTestConnection : IDisposable
         private readonly StreamReader _reader;
 
-        protected StreamBackedTestConnection(Stream stream)
+        protected StreamBackedTestConnection(Stream stream, Encoding encoding = null)
```
Default parameters must be a compile time constant.
So (unfortunately) this won't work that way.

JunTaoLuo replied:
Ah right, well the thought made more sense when reading code on Github 😄

JunTaoLuo reviewed (approved):
Will circle back on reviewing tests. But src changes LGTM

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http3/Http3FrameWriter.cs:
```diff
@@ -279,15 +280,15 @@ public ValueTask<FlushResult> WriteResponseTrailersAsync(long streamId, HttpResp
+                catch (Exception ex) // Any encoding error can corrupt connection state.
                 {
                     _log.QPackEncodingError(_connectionId, streamId, ex);
                     _http3Stream.Abort(new ConnectionAbortedException(ex.Message, ex), Http3ErrorCode.InternalError);
```
Should we be aborting the entire connection instead of just the stream?

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http3/Http3HeadersEnumerator.cs:
```diff
@@ -0,0 +1,152 @@
+
+        public Func<string, Encoding?> EncodingSelector { get; set; } = KestrelServerOptions.DefaultHeaderEncodingSelector;
+
+        public int QPackStaticTableId => GetResponseHeaderStaticTableId(_knownHeaderType);
```
Where/how is this used?

Tratcher replied:
The pattern is copied from Http2HeadersEnumerator and will be used when implementing https://github.com/dotnet/aspnetcore/issues/33980.

halter73 commented on src/Servers/Kestrel/Core/test/HttpResponseHeadersTests.cs:
```diff
@@ -129,6 +130,117 @@ public void AddingControlOrNonAsciiCharactersToHeadersThrows(string key, string
+        [InlineData("Da\u0080ta")]
+        [InlineData("Da™ta")]
+        [InlineData("Dašta")]
+        public void AddingControlOrNonAsciiCharactersToHeaderPropertyThrows(string value)
```
👍

Tratcher committed:
Revert stackalloc change

Tratcher committed:
H/Qpack encoding errors

Tratcher committed:
Fix test issues

Tratcher commented on src/Servers/Kestrel/Core/test/PipelineExtensionTests.cs:
```diff
@@ -148,8 +148,8 @@ public void WriteAscii()
         [Theory]
-        [InlineData(2, 1)]
-        [InlineData(3, 1)]
+        [InlineData(2, 2)]
+        [InlineData(3, 2)]
```
We're now using GetMaxByteCount(1) to ensure we have enough space to write at least one character into the buffer, but for ASCII this returns 2 for some reason. I don't know how one char could become two ASCII bytes.

Tratcher commented:
Rebased to resolve merge conflicts. I fixed a bug in WriteEncodedMultiWrite uncovered by the new benchmarks where it didn't ensure buffer space the first time through the loop.

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Http3/Http3HeadersEnumerator.cs:
```diff
@@ -0,0 +1,152 @@
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http3
+{
+    internal sealed class Http3HeadersEnumerator : IEnumerator<KeyValuePair<string, string>>
```
Nice. I had plans to do this 😄

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Infrastructure/IKestrelTrace.cs:
```diff
@@ -72,7 +72,7 @@ internal interface IKestrelTrace : ILogger
         void HPackDecodingError(string connectionId, int streamId, HPackDecodingException ex);
 
-        void HPackEncodingError(string connectionId, int streamId, HPackEncodingException ex);
+        void HPackEncodingError(string connectionId, int streamId, Exception ex);
```
If `HPackEncodingError` takes `Exception` then should `HPackDecodingError` (and QPack equivalent) be updated to match? I like the consistency.

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Http2/HPackHeaderWriter.cs:
```diff
@@ -99,6 +99,9 @@ private static bool EncodeHeadersCore(DynamicHPackEncoder hpackEncoder, Http2Hea
                 var staticTableId = headersEnumerator.HPackStaticTableId;
                 var name = headersEnumerator.Current.Key;
                 var value = headersEnumerator.Current.Value;
+                var valueEncoding =
+                    ReferenceEquals(headersEnumerator.EncodingSelector, KestrelServerOptions.DefaultHeaderEncodingSelector)
+                    ? null : headersEnumerator.EncodingSelector(name);
```
Couldn't this check happen once when the header enumerator is initialized? i.e. only set an encoding selector if it isn't equal to DefaultHeaderEncodingSelector

JamesNK replied:
Although... the performance difference between `ReferenceEquals` and `!= null` is probably zero 😬

Tratcher replied:
Yeah I don't think it would be a performance difference, at best it would make the code more concise and at worst it might make it easier to miss a code path where we forget to initialize it. I'll leave it for now since it's consistent with how we do it for request headers.

halter73 replied:
I was thinking about this same thing, but I was in no position to talk since I also did it this way for request headers 😅

Tratcher committed:
Test fix

Tratcher committed:
Log symmetry

Tratcher commented:
/azp run

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http3/Http3FrameWriter.cs:
```diff
@@ -280,9 +281,12 @@ public ValueTask<FlushResult> WriteResponseTrailersAsync(long streamId, HttpResp
                 {
                     _log.QPackEncodingError(_connectionId, streamId, ex);
+                    _connectionContext.Abort(new ConnectionAbortedException(ex.Message, ex));
                     _http3Stream.Abort(new ConnectionAbortedException(ex.Message, ex), Http3ErrorCode.InternalError);
```
Shouldn't aborting the `_connectionContext` abort the stream implicitly?

Tratcher replied:
@JamesNK I wasn't sure about that part, I know state is tracked much differently from HTTP/2.

davidfowl commented:
Please watch the benchmarks after this change makes it all the way in.
